### ISSUE TITLE
Unordered empty VTT cues affects real cues

### ISSUE THREAD

**Issue by jakubvojacek** [shaka-project/shaka-packager#609]
## System info
**Operating System**: latest debian
**Shaka Packager Version**: 5d808959

## Issue and steps to reproduce the problem
I have downloaded the Sintel trailer (https://download.blender.org/durian/trailer/) and packaged it together with some random subtitles below. The catch with the subtitles is that they are unordered with some empty cues. We use empty cues for live content in order for shaka packager to keep producing subtitle chunks. That is why we're printing empty cue every second (we use much smaller chunks for live content than for VODs). They are unordered because we can only know when the real cue is at the end of the cue, not the beginning (and there are some more delays because of OCR / translations...). We cannot guarantee that the correct order while maintaining empty cues every second.

```
WEBVTT

STYLE

00:00:09.448 --> 00:00:09.449

00:00:12.448 --> 00:00:12.449

00:00:15.288 --> 00:00:15.289

00:00:18.168 --> 00:00:18.169

00:00:12.946 --> 00:00:17.186
naší bolesti.
-Ano, pane. Hned to bude, pane.

00:00:21.048 --> 00:00:21.049

00:00:23.928 --> 00:00:23.929

00:00:17.187 --> 00:00:20.339
-Oh! Je to moc vysoko,
Taťko Šmoulo.

00:00:26.728 --> 00:00:26.729

00:00:20.340 --> 00:00:25.788
-Nešiko, šlapeš mi po hlavě.
-Jé! Promiň, Koumáku! Je to lepší?

00:00:29.648 --> 00:00:29.649

00:00:25.789 --> 00:00:28.871
-Eh!Je,au!
-Nenízač!

00:00:32.408 --> 00:00:32.409

00:00:35.368 --> 00:00:35.369

00:00:30.397 --> 00:00:33.447
-Držte se, Šmoulové!
-Snažím se, Tat'ko Smoulo!

00:00:38.168 --> 00:00:38.169

00:00:33.448 --> 00:00:35.992
-Oh! Au!

00:00:40.968 --> 00:00:40.969

00:00:35.993 --> 00:00:38.827
-Proč mě radši král nepožádá,
abych mu vyvolal ducha

00:00:43.768 --> 00:00:43.769

00:00:38.828 --> 00:00:41.679
nebo nějaké zaříkadlo?
Ale připravit lektvar
```

**Packager Command**:
```
packager  'in=/data/assets/vods/14/sintel-cmaf-test-1446/sintel-trailer-720p-mp4_transcoded.mp4,stream=0,bandwidth=400000,init_segment=video/sintel-trailer-720p-mp4_video_0.mp4,segment_template=video/0/video_0_sintel-trailer-720p-mp4-$Number$.mp4' 'in=/data/assets/vods/14/sintel-cmaf-test-1446/sintel-trailer-720p-mp4_transcoded.mp4,stream=1,bandwidth=1000000,init_segment=video/sintel-trailer-720p-mp4_video_1.mp4,segment_template=video/1/video_1_sintel-trailer-720p-mp4-$Number$.mp4' 'in=/data/assets/vods/14/sintel-cmaf-test-1446/sintel-trailer-720p-mp4_transcoded.mp4,stream=2,bandwidth=128000,language=eng,init_segment=audio/sintel-trailer-720p-mp4_audio_eng_stereo.mp4,segment_template=audio/eng/audio_eng_stereo_sintel-trailer-720p-mp4-$Number$.mp4' 'in=/data/assets/vods/14/sintel-cmaf-test-1446/5d209b41ef19d-test.vtt.vtt,stream=text,init_segment=text/5d209b41ef19d-test.vtt.vtt_init.mp4,segment_template=text/cs/5d209b41ef19d-test.vtt.vtt_$Number$.m4s,language=cs' --mpd_output sintel-trailer-720p-mp4.mpd \
--generate_static_mpd \
--fragment_duration 9.6 \
--segment_duration 9.6 \
--default_language=eng  \
--default_text_language=xxx \
--allow_approximate_segment_timeline
```

Extra steps to reproduce the problem?
(1) the packaged VOD is available here https://edge1.motv.eu///vods/14/sintel-cmaf-test-1446/sintel-trailer-720p-mp4.mpd

**What is the expected result?**
Based on @kqyang comment here https://github.com/google/shaka-packager/issues/380#issuecomment-405725038, I was hoping that the empty cues would be completely stripped and the real cues would be displayed as per their timings. 

**What happens instead?**
The real cues are being cut from both sides. For example the cue starting `00:00:35.993 --> 00:00:38.827` is displayed at second 37 instead of 35. We need this for live content, I just used the VOD as an example - for VOD there is no need for empty cues.

Is there any way how to make this work according to our expectations, please?

Thank you
Jakub

**Comment by kqyang** [comment]
@jakubvojacek That is because the timing in the text cues (including empty cues) is used to determine whether a new text chunk needs to be generated. After the new text chunk is generated, there is no way to put additional data inside. So the text cues should be strictly in order.

> We cannot guarantee that the correct order while maintaining empty cues every second.

We'll need to find another solution if we cannot guarantee the order. I have some thoughts but we don't have resource and bandwidth to work on it right now. 

Would it be possible for you to have a better way of generating empty cues which takes the timing of latest actual cue into consideration? This should be do-able as you can read the file before generating the empty cue.

**Comment by jakubvojacek** [comment]
Hello,

> Would it be possible for you to have a better way of generating empty cues which takes the timing of latest actual cue into consideration? This should be do-able as you can read the file before generating the empty cue.

packager needs the empty cue in order for the segment to be generated, right? That is why we need empty cues every 1 second. The subtitle extractor tool works in a way that it reports about the subtitle only after the cue has ended (since that's how some type of subs works, we have been told so by the maintainers), not when it begins. If you have a cue start at 0:10 and ending at 0:15, we will get notification about its existence at 0:15. We do not see the future, therefore, we couldn't have known at 0:10, 0:11, 0:12, 0:13, 0:14 that there is going to be cue at 0:15. If we wouldn't generate empty cues  0:10, 0:11, 0:12, 0:13, 0:14, there would be no segments. If we generate them, the real cue gets cut off and is displayed for a second only.

Since you cannot work on it, we can overcome this with another script in between that will act sort like a timed buffer and only release captions after 5 (or more) seconds (and dropping empty if necessary) but that's adding more complexity and more things can go wrong

Thank you
Jakub

**Comment by kqyang** [comment]
Let's keep it open for discussion and track potential future work.

> If you have a cue start at 0:10 and ending at 0:15, we will get notification about its existence at 0:15. 

Do you mean that the cue will only be written and send to packager at 0:15? 

If that is the case, the only option would be to buffer and hold the segment, which means that there will be at least 5 seconds additional delay, regardless of whether it is handled inside packager or outside of packager.

Would it be possible to break up the segment in ccextractor then, for example, break the cue at [0:10, 0:15] to 5 cues with one second each? They can use the same id to signal that they are the same cue.

**Comment by jakubvojacek** [comment]
> Do you mean that the cue will only be written and send to packager at 0:15?

yes. I agree about the buffer, was just hoping that it would be commenting out one if condition in packager (something like deleting `continue if cueTimestamp < lastCueTimestamp`) instead of writing the buffer myself (already got it working anyway)

> Would it be possible to break up the segment in ccextractor then, for example, break the cue at [0:10, 0:15] to 5 cues with one second each? Not sure how it behaves in the player.

I can try to ask them but I think it would result with "blinking" subtitles on the screen in some players.

It would be best if this could be managed directly in packager instead of relying on 3rd party tools. For example

- if subtitle file is provided to packager, packager will by himself create empty chunks with the timestamp of the first video stream.
- packager will keep checking provided file reading it indefinitely if a past cue is found, empty chunk will be overwritten, if future cue is found, its added to buffer and written when the time is right

 But I understand that the use case of live subtitles is quite rare and nobody uses that but us

Thank you
Jakub

**Comment by kqyang** [comment]
> I can try to ask them but I think it would result with "blinking" subtitles on the screen in some players.

I think it worth trying as it improves the latency.

> It would be best if this could be managed directly in packager instead of relying on 3rd party tools.

That is our goal too. We just don't have enough resource right now.
